---
title: "Plot counts from transcriptomic data"
output:
  md_document:
    variant: gfm
---

# Plot counts from transcriptomic data

```{r packages, include=FALSE, message=FALSE}
for( package in c("knitr", "rprojroot", "tidyverse", "DESeq2", "biovisr")) {
  library(package, character.only = TRUE)
}
```

```{r knitr_options, include=FALSE, message=FALSE}
opts_chunk$set(fig.width=6, fig.height=4, include=TRUE, echo=TRUE)
```

```{r root-path, include=FALSE}
#set.seed(193746)
root_path <- find_root(is_rstudio_project)
Sys.setenv(ROOT_PATH = root_path)
```

```{r create-test-data, include=FALSE}
# test data for graph_counts_by_group_facet.R
# makes test_samples.tsv, rnaseq_adjusted_pvalues.tsv, 
# test_rnaseq_data.tsv and test_gene_metadata.tsv
set.seed(583)
num_rows <- 10
tmp <- makeExampleDESeqDataSet(n = num_rows, m = 12, betaSD = 2)
tmp2 <- makeExampleDESeqDataSet(n = num_rows, m = 12, betaSD = 2, interceptMean = 3)
# adjust sample names
rownames(colData(tmp2)) <- paste0('sample', 13:24)
# create samples file
df <- rbind(colData(tmp), colData(tmp2))
set.seed(714)
samples_df <- data.frame(
  sample = rownames(df),
  condition = rep( rep(c('wt', 'mut'), each = 6), 2 ),
  treatment = rep(c('control', 'treated'), each = 12),
  sex = sample(c("F", "M"), 24, replace = TRUE)
)
write_tsv(samples_df, file = file.path(root_path, 'test_data', 'test_samples.tsv'))

tmp <- DESeq(tmp)
res <- results(tmp)
tmp2 <- DESeq(tmp2)
# create gene info
set.seed(208)
starts <- sample(1:10000, num_rows)
test_all_data <- tibble(
  'GeneID' = sprintf('ENSTEST%03d', seq_len(num_rows)),
  'chr' = sample(1:25, num_rows, replace = TRUE),
  'start' = starts,
  'end' = as.integer(starts + 100),
  'strand' = sample(c('1', '-1'), num_rows, replace = TRUE),
  'p value' = res$pvalue,
  'Adjusted p value' = res$padj,
  'Gene name' = paste0('gene-', seq_len(num_rows)),
  'Class' = c(sample(c('TF', 'DNAPol', 'GPCR'), num_rows - 2, replace = TRUE), NA, NA),
  'GO_BP' = c(rep(c('Translation', 'UPR'), each = 4), 'Immune Response', 'Immune Response'),
  'GO_CC' = c('Cytoplasm', rep('Nucleus', 5), rep('Cytoplasm', 4)),
  'GO_MF' = c(rep(c('Kinase activity', 'ATPase activity'), 5))
)
test_all_data$chr <- factor(test_all_data$chr, levels = unique(test_all_data$chr))
test_all_data$strand <- factor(test_all_data$strand)

# create pvalues file
test_all_data |> 
  mutate(condition1 = "wt",
         condition2 = "mut",
         log2fc = res$log2FoldChange) |> 
  dplyr::select(GeneID, condition1, condition2, adjp = `Adjusted p value`, log2fc) |> 
  write_tsv(file = file.path(root_path, 'test_data', 'rnaseq_adjusted_pvalues.tsv'))

# create counts file
counts_1 <- counts(tmp)
counts_2 <- counts(tmp2)
colnames(counts_1) <- sub("$", " count", colnames(counts_1))
colnames(counts_2) <- sub("$", " count", colnames(counts_2))
norm_counts_1 <- counts(tmp, normalized = TRUE)
norm_counts_2 <- counts(tmp2, normalized = TRUE)
colnames(norm_counts_1) <- sub("$", " normalised count", colnames(norm_counts_1))
colnames(norm_counts_2) <- sub("$", " normalised count", colnames(norm_counts_2))
test_rnaseq_data <- cbind(
  test_all_data,
  counts_1,
  counts_2,
  norm_counts_1,
  norm_counts_2
)
write_tsv(test_rnaseq_data, file = file.path(root_path, 'test_data', 'test_rnaseq_data.tsv'))

# subset to 3 genes
test_all_data %>% filter(., GeneID %in% c('ENSTEST005', 'ENSTEST006', 'ENSTEST009')) %>% 
  select(., GeneID, Name = `Gene name`) %>% 
  write_tsv(., file = file.path(root_path, 'test_data', 'test_genes_to_label.txt'))

# make gene metadata file
gene_metadata <- select(test_all_data, GeneID, GO_BP, GO_CC, GO_MF) %>% 
  pivot_longer(., -GeneID, names_to = 'category') %>% 
  filter(., !is.na(value)) %>% 
  arrange(., category, value)

write_tsv(gene_metadata, file = file.path(root_path, 'test_data', 'test_gene_metadata.tsv'))
```

[Go to script](../../graph_counts_by_group_facet.R)

This is a script to produce counts plots from RNA-seq (or DETCT) data.  
For information on all the options run

    Rscript graph_counts_by_group_facet.R --help

The required arguments are a samples file and a sig file.
It expects the sample file to have a header with the column names.
It expects one of the columns to be called "sample".
e.g.

    sample  condition
    sample_1    wt
    sample_2    mut

There is an example samples file and sig file in the test_data directory of
this repository.
The simplest way to run the script would be this:

```{bash counts-basic, eval=FALSE, echo=TRUE}
Rscript graph_counts_by_group_facet.R \
test_data/test_samples.tsv test_data/test_rnaseq_data.tsv
```

```{bash run-counts-basic, echo=FALSE, results='hide'}
echo "ENSTEST001" > test_gene.txt

Rscript $ROOT_PATH/graph_counts_by_group_facet.R \
--output_file test_counts_basic.png \
--genes_file test_gene.txt \
--width 480 \
--height 360 \
$ROOT_PATH/test_data/test_samples.tsv \
$ROOT_PATH/test_data/test_rnaseq_data.tsv
```

![Basic count plot showing the normalised counts for the wt condition as blue
circles and the mutant condition as orange circles](test_counts_basic.1.png "Basic count plot")

By default it tries to use a column named condition in the samples file as the
x variable. The default is to colour the points by condition as well.

The \--x_variable option allows you to name a column to use as the x axis and
you can also specify columns to use for colours (\--colour_variable) and
shapes (\--shape_variable).
The \--crossbar option plots a bar to show the mean/median of each group.
It is also possible to supply a file of Ensembl gene ids and the script will
only make plots for those gene ids in the data.

The example below uses almost all the available options

```{bash counts-all-options, eval=FALSE, echo=TRUE}
# create a test ids file
echo "ENSTEST006" > test_gene.txt

graph_counts_by_group_facet.R \
--output_file test-condition-sex-treatment.pdf \
--genes_file test_gene.txt \
--x_variable condition \
--colour_variable condition \
--colour_palette 'wt=#0000ff,mut=#ff0000' \
--shape_variable sex \
--facet_variable treatment \
--crossbar median \
--width 12 \
--height 8 \
--theme_base_size 14 \
--rotate_xaxis_labels \
--seed 7635 \
--no_pvalue \
test_samples.tsv test_rnaseq_data.tsv
```

```{bash run-counts-all-options, echo=FALSE, results='hide'}
echo "ENSTEST006" > test_gene.txt

Rscript $ROOT_PATH/graph_counts_by_group_facet.R \
--output_file count_plot_everything.png \
--genes_file test_gene.txt \
--x_variable condition \
--colour_variable condition \
--colour_palette 'wt=#0000ff,mut=#ff0000' \
--shape_variable sex \
--facet_variable treatment \
--crossbar median \
--width 480 \
--height 360 \
--theme_base_size 14 \
--rotate_xaxis_labels \
--seed 7635 \
--no_pvalue \
$ROOT_PATH/test_data/test_samples.tsv $ROOT_PATH/test_data/test_rnaseq_data.tsv
```

![Count plot showing the normalised counts for the wt condition in blue and the
mutant condition in red. The points are split by the control or treated. Sex is
displayed as circle for female and squares for male](count_plot_everything.1.png "Count plot by condition by treatment")

There is also an option to supply a file of p-values which be added to the pot with a line from one condition to another. The file should look like this:

|GeneID|condition1|condition2|adjp|log2fc|
|:---|:---|:---|---:|---:|
|ENSTEST001|wt|mut|6.52e-5|3.378|

The `log2fc` column is optional. The \--asterisks option converts the p-values to asterisks.

```{bash counts-pval, eval=FALSE, echo=TRUE}
Rscript graph_counts_by_group_facet.R \
--pvalue_file rnaseq_adjusted_pvalues.tsv --asterisks \
test_data/test_samples.tsv test_data/test_rnaseq_data.tsv
```

```{bash run-counts-pval, echo=FALSE, results='hide'}
Rscript $ROOT_PATH/graph_counts_by_group_facet.R \
--output_file test_counts_pval.png \
--genes_file test_gene.txt \
--pvalue_file $ROOT_PATH/test_data/rnaseq_adjusted_pvalues.tsv --asterisks \
--width 480 \
--height 360 \
$ROOT_PATH/test_data/test_samples.tsv \
$ROOT_PATH/test_data/test_rnaseq_data.tsv
```

![Count plot showing the normalised counts for the wt condition as blue
circles and the mutant condition as orange circles. A line with three asterisks above it, shows that the difference between the two groups is significant.](test_counts_pval.1.png "Count plot by condition with p-values")

The other options are:

- \--log10 - Use a log10 scaled y-axis
- \--output_data_file Output an Rdata file of the plot objects
- \--no_jitter - removes the jitter from the points
- \--seed - random seed to make the jitter reproducible
- \--no_pvalue Don't add a pvalue to the plot title
- \--detct - input data is DeTCT rather than RNAseq

## Required packages

- [optparse](https://github.com/trevorld/r-optparse)
- [tidyverse](https://www.tidyverse.org/)
- [biovisr](https://github.com/richysix/biovisr)
- [rnaseqtools](https://github.com/richysix/rnaseqtools)

