---
title: "Gene expression heatmaps"
author: "Richard White"
date: "2023-05-04"
output:
  md_document:
    variant: gfm
---

# Create gene expression heatmaps from a transcriptomics experiment

```{r packages, include=FALSE, message=FALSE}
for( package in c("knitr", "rprojroot", "DESeq2", "tidyverse")) {
  library(package, character.only = TRUE)
}
```

```{r knitr_options, include=FALSE, message=FALSE}
opts_chunk$set(include=TRUE, echo=TRUE)
```

```{r root-path, include=FALSE}
root_path <- find_root(is_rstudio_project)
Sys.setenv(ROOT_PATH = root_path)
```

```{r create-test-data, include=FALSE}
# Required test file are
# test_samples.tsv
# test_rnaseq_data.tsv
# test_genes.txt
# test_samples_long.tsv
# test_samples.tsv and test_rnaseq_data.tsv are created by docs/graph_counts_by_group_facet/graph_counts_by_group_facet.Rmd

# read in test_samples.tsv
samples_df <- read_tsv(file.path(root_path, 'test_data', 'test_samples.tsv'))
# pivot samples file to use as metadata
samples_df %>% 
  pivot_longer(., cols = condition:sex, names_to = "category",
               values_to = "value") |> 
  write_tsv(file = file.path(root_path, 'test_data', 'test_samples_long.tsv'))
```

## gene_expr_heatmap.R

Script to produce a heatmap from RNAseq data. It expects a sample file and a count file (e.g. sig.tsv)
and the name of the output image file.

There is an example samples file and sig file in the test_data directory of this repository.
For a basic heatmap do:

```{bash heatmap, eval=FALSE, echo=TRUE}
Rscript gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster rows \
--colour_palette magma --cell_colour grey80 \
--gene_names --sample_names \
test_data/test_samples.tsv test_data/test_rnaseq_data.tsv test_data/test_heatmap.pdf
```

```{bash run-heatmap, echo=FALSE, results='hide'}
Rscript $ROOT_PATH/gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster rows \
--colour_palette magma --cell_colour grey80 \
--gene_names --sample_names \
--height 300 \
$ROOT_PATH/test_data/test_samples.tsv \
$ROOT_PATH/test_data/test_rnaseq_data.tsv \
rnaseq-heatmap.png
```

The available fill_palettes are those from [viridis](https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html) and
[ColorBrewer](https://colorbrewer2.org/) via [scale_fill_distiller](https://ggplot2.tidyverse.org/reference/scale_brewer.html)

![Gene expression heatmap. Genes are displayed in rows with the samples in the
columns. Each box is coloured according to the expression of the gene/sample
combination](rnaseq-heatmap.png "RNAseq heatmap")

It is also possible to supply a list of gene ids to subset the heatmap to.

```{bash heatmap-subset, eval=FALSE, echo=TRUE}
# create a test ids file
echo -e "GeneID\tName
ENSTEST005\tgene-5
ENSTEST006\tgene-6
ENSTEST009\tgene-9" > test_genes.txt

Rscript gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster rows \
--colour_palette magma --cell_colour grey80 \
--gene_names --sample_names \
--genes_file test_genes.txt \
test_samples.tsv test_rnaseq_data.tsv test_heatmap_subset.pdf
```

```{bash run-heatmap-subset, echo=FALSE, results='hide'}
echo -e "GeneID\tName
ENSTEST005\tgene-5
ENSTEST006\tgene-6
ENSTEST009\tgene-9" > test_genes.txt

Rscript $ROOT_PATH/gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster rows \
--colour_palette magma --cell_colour grey80 \
--gene_names --sample_names \
--genes_file test_genes.txt \
--height 180 \
$ROOT_PATH/test_data/test_samples.tsv \
$ROOT_PATH/test_data/test_rnaseq_data.tsv \
rnaseq-heatmap-subset.png
```

![Gene expression heatmap. Genes are displayed in rows with the samples in the
columns. Each box is coloured according to the expression of the gene/sample
combination. Only three of the original rows are shown](rnaseq-heatmap-subset.png "RNAseq heatmap subset")

Or only label the supplied genes

```{bash heatmap-subset-labels, eval=FALSE, echo=TRUE}
Rscript gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster rows \
--colour_palette magma --cell_colour grey80 \
--sample_names \
--genes_to_label test_genes.txt \
test_samples.tsv test_rnaseq_data.tsv test_heatmap_subset_labels.pdf
```

```{bash run-heatmap-subset-labels, echo=FALSE, results='hide'}
Rscript $ROOT_PATH/gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster rows \
--colour_palette magma --cell_colour grey80 \
--sample_names \
--genes_to_label test_genes.txt \
--height 200 \
$ROOT_PATH/test_data/test_samples.tsv \
$ROOT_PATH/test_data/test_rnaseq_data.tsv \
rnaseq-heatmap-subset-labels.png
```

![Gene expression heatmap. Genes are displayed in rows with the samples in the
columns. Each box is coloured according to the expression of the gene/sample
combination. Only three of the rows are labelled](rnaseq-heatmap-subset-labels.png "RNAseq heatmap with a subset of genes labelled")

A file of sample metadata can also be supplied and will be plotted as a heatmap
under the expression heatmap.
```{bash heatmap-sample-metadata, eval=FALSE, echo=TRUE}
Rscript gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster rows \
--colour_palette magma --cell_colour grey80 \
--gene_names --sample_names \
--sample_metadata_file test_samples_long.tsv \
--sample_metadata_ycol category \
--sample_metadata_fill value \
--relative_heights "0,8,2" \
test_samples.tsv test_rnaseq_data.tsv test_heatmap_with_sample_metadata.pdf
```

```{bash run-heatmap-sample-metadata, echo=FALSE, results='hide'}
Rscript $ROOT_PATH/gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster rows \
--colour_palette magma --cell_colour grey80 \
--gene_names --sample_names \
--sample_metadata_file $ROOT_PATH/test_data/test_samples_long.tsv \
--sample_metadata_ycol category \
--sample_metadata_fill value \
--relative_heights "0,8,2" \
--height 360 \
$ROOT_PATH/test_data/test_samples.tsv \
$ROOT_PATH/test_data/test_rnaseq_data.tsv \
rnaseq-heatmap-sample-metadata.png
```

![Gene expression heatmap with sample metadata heatmap. Genes are displayed in
rows with the samples in the columns. Each box is coloured according to the
expression of the gene/sample combination. A second heatmap shows the metadata
associated with each sample](rnaseq-heatmap-sample-metadata.png "RNAseq heatmap with metadata")

The `--gene_tree` and `--sample_tree` options plot trees for the gene/sample clustering.

```{bash heatmap-trees, eval=FALSE, echo=TRUE}
Rscript gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster both \
--colour_palette magma --cell_colour grey80 \
--gene_tree --sample_tree --gene_names --sample_names \
--sample_metadata_file test_samples_long.tsv \
--sample_metadata_ycol category \
--sample_metadata_fill value \
--relative_heights 1,3,1 --relative_widths 1,3 \
test_samples.tsv test_rnaseq_data.tsv test_heatmap_with_trees.pdf
```

```{bash run-heatmap-trees, echo=FALSE, results='hide'}
Rscript $ROOT_PATH/gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster both \
--colour_palette magma --cell_colour grey80 \
--gene_tree --sample_tree --gene_names --sample_names \
--sample_metadata_file $ROOT_PATH/test_data/test_samples_long.tsv \
--sample_metadata_ycol category \
--sample_metadata_fill value \
--relative_heights 1,3,1 --relative_widths 1,3 \
--height 480 \
$ROOT_PATH/test_data/test_samples.tsv \
$ROOT_PATH/test_data/test_rnaseq_data.tsv \
rnaseq-heatmap-trees.png
```

![Gene expression heatmap with sample metadata heatmap and clustering trees.
Genes are displayed in rows with the samples in the columns. Each box is
coloured according to the expression of the gene/sample combination.
A second heatmap shows the metadata associated with each sample and the
clustering trees for the genes and samples are shown above and to the
left of the main expression heatmap](rnaseq-heatmap-trees.png "RNAseq heatmap with sample metadata and clustering trees")

A file of gene metadata can also be supplied and will be plotted as a heatmap
to the right of the expression heatmap.

```{bash heatmap-gene-metadata, eval=FALSE, echo=TRUE}
Rscript gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster both \
--colour_palette magma --cell_colour grey80 \
--gene_tree --sample_tree --gene_names --sample_names \
--sample_metadata_file test_samples_long.tsv \
--sample_metadata_ycol category \
--sample_metadata_fill value \
--gene_metadata_file test_gene_metadata.tsv \
--gene_metadata_xcol category --gene_metadata_fill value \
--gene_metadata_fill_palette cividis \
--relative_heights 1,3,1 --relative_widths 1,3,1 \
--width 640 --height 480 \
test_samples.tsv test_rnaseq_data.tsv test_heatmap_with_gene-metadata.pdf
```

```{bash run-heatmap-gene-metadata, echo=FALSE, results='hide'}
Rscript $ROOT_PATH/gene_expr_heatmap.R --transform rlog \
--center_and_scale --cluster both \
--colour_palette magma --cell_colour grey80 \
--gene_tree --sample_tree --gene_names --sample_names \
--sample_metadata_file $ROOT_PATH/test_data/test_samples_long.tsv \
--sample_metadata_ycol category \
--sample_metadata_fill value \
--gene_metadata_file $ROOT_PATH/test_data/test_gene_metadata.tsv \
--gene_metadata_xcol category --gene_metadata_fill value \
--gene_metadata_fill_palette Dark2 \
--relative_heights 1,3,1 --relative_widths 1,3,1 \
--width 640 --height 480 \
$ROOT_PATH/test_data/test_samples.tsv \
$ROOT_PATH/test_data/test_rnaseq_data.tsv \
rnaseq-heatmap-gene-metadata.png
```

![Gene expression heatmap with gene and sample metadata heatmaps and clustering trees.
Genes are displayed in rows with the samples in the columns. Each box is coloured
according to the expression of the gene/sample combination. Two more heatmaps show the
metadata associated with each gene and sample and the clustering trees for the genes and
samples are shown above and to the left of the main expression heatmap](rnaseq-heatmap-gene-metadata.png "RNAseq heatmap with gene and sample metadata and clustering trees")

## Required packages

* [tidyverse](https://www.tidyverse.org/)
* [viridis](https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html)
* [ggrepel](https://cran.r-project.org/web/packages/ggrepel/vignettes/ggrepel.html)
* [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)
* [patchwork](https://patchwork.data-imaginist.com/)
* [biovisr](https://github.com/richysix/biovisr)
* [rnaseqtools](https://github.com/richysix/rnaseqtools)

